### PR TITLE
Graduate Evented PLEG to Beta

### PR DESCRIPTION
**Pull request by harche**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Graduate `Evented PLEG` feature to Beta

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/enhancements/issues/3386

#### Special notes for your reviewer:
As a part of graduating the `Evented PLEG` to Beta we had decided to add 4 new grpc connection related [metrics](https://github.com/kubernetes/enhancements/pull/3817/files#diff-4a48c7915d4769b7522c1c9711a927f4ebacfcefe8b2b60c9edf6ffd43e35498R421). However, while implementing I realized that we don't have to implement propsed [evented_pleg_notifications_received](https://github.com/kubernetes/enhancements/pull/3817/files#diff-4a48c7915d4769b7522c1c9711a927f4ebacfcefe8b2b60c9edf6ffd43e35498R424) metric because as a part of implementing [evented_pleg_connection_latency_seconds](https://github.com/kubernetes/enhancements/pull/3817/files#diff-4a48c7915d4769b7522c1c9711a927f4ebacfcefe8b2b60c9edf6ffd43e35498R423), a histogram metric, we automatically get a metric `kubelet_evented_pleg_connection_latency_seconds_count` that is capturing the total count of the events. 

```
# curl -k --cert /var/run/kubernetes/client-admin.crt --key /var/run/kubernetes/client-admin.key https://localhost:10250/metrics | grep -i evented_pleg_
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# HELP kubelet_evented_pleg_connection_error_count [ALPHA] The number of errors encountered during the establishment of streaming connection with the CRI runtime.
# TYPE kubelet_evented_pleg_connection_error_count counter
kubelet_evented_pleg_connection_error_count 0
# HELP kubelet_evented_pleg_connection_latency_seconds [ALPHA] The latency of streaming connection with the CRI runtime, measured in seconds.
# TYPE kubelet_evented_pleg_connection_latency_seconds histogram
kubelet_evented_pleg_connection_latency_seconds_bucket{le="0.005"} 14
kubelet_evented_pleg_connection_latency_seconds_bucket{le="0.01"} 14
kubelet_evented_pleg_connection_latency_seconds_bucket{le="0.025"} 14
kubelet_evented_pleg_connection_latency_seconds_bucket{le="0.05"} 14
kubelet_evented_pleg_connection_latency_seconds_bucket{le="0.1"} 14
kubelet_evented_pleg_connection_latency_seconds_bucket{le="0.25"} 14
kubelet_evented_pleg_connection_latency_seconds_bucket{le="0.5"} 14
kubelet_evented_pleg_connection_latency_seconds_bucket{le="1"} 14
kubelet_evented_pleg_connection_latency_seconds_bucket{le="2.5"} 14
kubelet_evented_pleg_connection_latency_seconds_bucket{le="5"} 14
kubelet_evented_pleg_connection_latency_seconds_bucket{le="10"} 14
kubelet_evented_pleg_connection_latency_seconds_bucket{le="+Inf"} 14
kubelet_evented_pleg_connection_latency_seconds_sum 0.0044651019999999994
kubelet_evented_pleg_connection_latency_seconds_count 14
# HELP kubelet_evented_pleg_connection_success_count [ALPHA] The number of times the connection was successfully established with the CRI Runtime.
# TYPE kubelet_evented_pleg_connection_success_count counter
kubelet_evented_pleg_connection_success_count 1
100  130k    0  130k    0     0  11.0M      0 --:--:-- --:--:-- --:--:-- 11.5M
```

#### Does this PR introduce a user-facing change?

```release-note
Graduate CRI Events driven Pod LifeCycle Event Generator (Evented PLEG) to Beta
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3386
```

### CODE REVIEW

File: pkg/kubelet/metrics/metrics.go
Comment (saschagrunert) [inline pkg/kubelet/metrics/metrics.go:258]: I think we should clarify that it's the amount of successful `GetContainerEvents` RPC calls. The connection establishment with the runtime will be done once on Kubelet startup, so the wording seems misleading to me personally.

**Comment by sftim** [comment]
For the changelog entry, we might want to provide more context. Some people won't know what a PLEG is.

File: pkg/kubelet/metrics/metrics.go
Comment (harche) [reply pkg/kubelet/metrics/metrics.go:258]: @saschagrunert Thanks for that feedback. 

```
	// EventedPLEGConn is a Counter that tracks the number of times  
	// a streaming client was obtained to receive CRI Events. 
``` 
Does this sound better? Feel free to suggest if any changes.

Comment (saschagrunert) [reply pkg/kubelet/metrics/metrics.go:258]: Yep, sounds way better, thank you!

Comment (harche) [reply pkg/kubelet/metrics/metrics.go]: Thanks, updated.

**Comment by harche** [comment]
> For the changelog entry, we might want to provide more context. Some people won't know what a PLEG is.

Thanks for that feedback. I elaborated a bit more on what the `Evented PLEG` is about in the changelog.

**Review by saschagrunert — APPROVED** [review]
/lgtm

File: pkg/features/kube_features.go
Comment (pacoxu) [inline pkg/features/kube_features.go:929]: Could you handle this automatically by container runtime? I mean that kubelet can check if it is supported and uses evented-pleg if it is supported and falls back to using old behavior if it is not supported.

Comment (harche) [reply pkg/features/kube_features.go:929]: We already [handle this in the code](https://github.com/kubernetes/kubernetes/blob/d446bebca8008ff8ba8906c76647c0ae40dbb6f0/pkg/kubelet/pleg/evented.go#L181). However, we still want the user to explicitly set the feature gate here. This way the feature gate for this particular feature is respected and is only used when set by the user. 

The code I pointed out takes care of the situation where the user has enabled this feature gate but for some reason the runtime is unable to send any event (either the runtime doesn't support evented pleg or there is some error in the runtime).

Comment (ruiwen-zhao) [reply pkg/features/kube_features.go:929]: With the default set to false, is the Beta promotion basically telling users that this feature is more mature?

**Comment by SergeyKanzhelev** [comment]
Was this graduation criteria completed?

> Beta
> - Add E2E Node Conformance presubmit job in CI
> - Add E2E Node Conformance periodic job in CI

**Comment by SergeyKanzhelev** [comment]
/hold

I do not see any PRs with tests linked to the KEP

**Comment by harche** [comment]
> /hold
> 
> I do not see any PRs with tests linked to the KEP

@SergeyKanzhelev The [presubmit](https://github.com/kubernetes/test-infra/pull/28366/files#diff-9495ba672589316a9265fd090ecb9c3c74fd414a11386acafda27df4bd409409R742) and [periodic](https://github.com/kubernetes/test-infra/pull/28592/files#diff-821d2874f55b26a240668fc5d17ae205f20dd433e61f0744f61a8ca72648a3dfR292) jobs were merged a while ago.

**Comment by harche** [comment]
Also, the test grid links have been added to the enhancement doc as well, https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3386-kubelet-evented-pleg/README.md#e2e-tests

**Comment by SergeyKanzhelev** [comment]
> https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3386-kubelet-evented-pleg/README.md#e2e-tests

My understanding of this graduation criteria was adding specific tests for EventPLEG - ones that were mentioned as a next step at a CI meeting: https://docs.google.com/document/d/1fb-ugvgdSVIkkuJ388_nhp2pBTy_4HEVg5848Xy7n5U/edit#bookmark=id.an5fifwuaef

Tests that the events are not lot under high load, tests with a high number of Pods and Containers, consistency tests that the current state is properly reflected by the kubelet.

Was my understanding wrong?

**Comment by harche** [comment]
> > https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3386-kubelet-evented-pleg/README.md#e2e-tests
> 
> My understanding of this graduation criteria was adding specific tests for EventPLEG - ones that were mentioned as a next step at a CI meeting: https://docs.google.com/document/d/1fb-ugvgdSVIkkuJ388_nhp2pBTy_4HEVg5848Xy7n5U/edit#bookmark=id.an5fifwuaef
> 
> Tests that the events are not lot under high load, tests with a high number of Pods and Containers, consistency tests that the current state is properly reflected by the kubelet.
> 
> Was my understanding wrong?

Those tests were never part of the criteria to graduate Evented PLEG to Beta.  The KEP, https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3386-kubelet-evented-pleg/README.md#beta, explicitly mentions the graduation criteria as running existing Node Conformance jobs (both presubmit and periodic) using Evented PLEG. 

```
Beta
Add E2E Node Conformance presubmit job in CI
Add E2E Node Conformance periodic job in CI
```

The aim of those those presubmit and periodic jobs is to make sure Evented PLEG conforms to the behaviour of existing Generic PLEG when it comes to generating PodLifeCycle Events.

**Comment by SergeyKanzhelev** [comment]
> Those tests (in that google doc) were never part of the criteria to graduate Evented PLEG to Beta. The KEP, https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3386-kubelet-evented-pleg/README.md#beta, explicitly mentions the graduation criteria as running existing Node Conformance jobs (both presubmit and periodic) using Evented PLEG.
> 
> ```
> Beta
> Add E2E Node Conformance presubmit job in CI
> Add E2E Node Conformance periodic job in CI
> ```
> 
> The aim of those those presubmit and periodic jobs is to make sure Evented PLEG conforms to the behaviour of existing Generic PLEG when it comes to generating PodLifeCycle Events.

Yes, It looks like I read what I wanted to read, not what is written. Sorry.

But the question stands. I think having extra tests will be very valuable for the feature. The question is whether we need it for Beta or can postpone for GA. I'd say it is needed for Beta. But can accept that it is ok to have them for Beta2 when it will be Enabled by default.

Can you please send update to enhancement text that will ensure we have those tests before making it default true as minimum?

**Comment by ruiwen-zhao** [comment]
> https://docs.google.com/document/d/1fb-ugvgdSVIkkuJ388_nhp2pBTy_4HEVg5848Xy7n5U/edit#bookmark=id.an5fifwuaef

Agree that we should have stress testing (with many pod/containers) are needed, because we might only be able to surface certain race conditions under load. I guess we can add that as a pre-req of making evented pleg default?

**Comment by harche** [comment]
@SergeyKanzhelev @ruiwen-zhao Thanks for your inputs. I will open a PR to update the enhancement to add those stress tests as pre-req before enabling it by default.

**Comment by harche** [comment]
> @SergeyKanzhelev @ruiwen-zhao Thanks for your inputs. I will open a PR to update the enhancement to add those stress tests as pre-req before enabling it by default.

https://github.com/kubernetes/enhancements/pull/3900

**Comment by harche** [comment]
@SergeyKanzhelev while we sort out [further steps](https://github.com/kubernetes/enhancements/pull/3900#discussion_r1129954722), do you still want to hold this PR?

**Comment by SergeyKanzhelev** [comment]
> @SergeyKanzhelev while we sort out [further steps](https://github.com/kubernetes/enhancements/pull/3900#discussion_r1129954722), do you still want to hold this PR?

I thought more about it and I think marking it beta would be OK for more people to try it out. I am still concerned with the feature stability, but since it's disabled by default.

@derekwaynecarr who is an approver for the KEP, what do you think?

**Comment by dchen1107** [comment]
I agreed with @SergeyKanzhelev at https://github.com/kubernetes/kubernetes/pull/115967#issuecomment-1457346939

But ok to delay the stress tests to GA (even I thought I requested it for beta for alpha) given the feature is off default. Also I noticed that the compatibility tests between Kubelet and container runtime are manual including some failover scenario in the KEP.

@harche Can you send a KEP enhancement PR to cover above two issues as GA blocker? Once the pr is ready, I can approve this one.

**Comment by harche** [comment]
> /lgtm /approve

Thanks @dchen1107 

@SergeyKanzhelev Do you still want to keep the hold?

**Pull request merged**
